docker/docker_client.go

```diff
@@ -221,6 +224,102 @@ func CheckAuth(ctx context.Context, sCtx *types.SystemContext, username, passwor
+// The limit is the max number of results desired
+// Note: The limit value doesn't work with all registries
+// for example registry.access.redhat.com returns all the results without limiting it to the limit value
+func SearchRegistry(ctx context.Context, sCtx *types.SystemContext, registry, image string, limit int) ([]SearchResult, error) {
+	type V2Results struct {
+		// Repositories holds the results returned by the /v2/_catalog endpoint
+		Repositories []string `json:"repositories"`
+	}
+	type V1Results struct {
+		// Results holds the results returned by the /v1/search endpoint
+		Results []SearchResult `json:"results"`
+	}
+	v2Res := &V2Results{}
+	v1Res := &V1Results{}
+
+	// The /v2/_catalog endpoint has been disabled for docker.io therefore the call made to that endpoint will fail
+	// So using the v1 hostname for docker.io for simplicity of implementation and the fact that it returns search results
+	if registry == dockerHostname {
+		registry = dockerV1Hostname
+	}
+
+	client, err := newDockerClientWithDetails(sCtx, registry, "", "", "", nil, "")
+	if err != nil {
+		return nil, errors.Wrapf(err, "error creating new docker client")
+	}
+
+	logrus.Debugf("trying to talk to v2 search endpoint\n")
+	resp, err := client.makeGETRequest(ctx, "/v2/_catalog", nil, nil)
+	if err == nil {
+		if err := json.NewDecoder(resp.Body).Decode(v2Res); err != nil {
+			return nil, err
+		}
+		searchRes := []SearchResult{}
+		for _, repo := range v2Res.Repositories {
+			if strings.Contains(repo, image) {
+				res := SearchResult{
+					Name: repo,
+				}
+				searchRes = append(searchRes, res)
+			}
+		}
+		return searchRes, nil
+	}
+	logrus.Debugf("error getting search results from v2 endpoint %q: %v", registry, err)
+
+	// set up the query values for the v1 endpoint
+	u := url.URL{
+		Path: "/v1/search",
+	}
+	q := u.Query()
+	q.Set("q", image)
+	q.Set("n", strconv.Itoa(limit))
+	u.RawQuery = q.Encode()
+
+	logrus.Debugf("trying to talk to v1 search endpoint\n")
+	resp, err = client.makeGETRequest(ctx, u.String(), nil, nil)
+	if err == nil {
+		if err := json.NewDecoder(resp.Body).Decode(v1Res); err != nil {
+			return nil, err
+		}
+		return v1Res.Results, nil
+	}
+	logrus.Debugf("error getting search results from v1 endpoint %q: %v", registry, err)
+
+	return nil, errors.Wrapf(err, "couldn't search registry %q", registry)
+}
+
+// makeGETRequest makes a get request and checks for the error and status code
+// if the status code is non 200 an error is returned
+func (c *dockerClient) makeGETRequest(ctx context.Context, path string, headers map[string][]string, stream io.Reader) (*http.Response, error) {
+	resp, err := c.makeRequest(ctx, "GET", path, headers, stream)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error making request")
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.Errorf("not OK, status code %q", resp.StatusCode)
```

Comment:
(Would this be better with `client.HandleErrorResponse`? Maybe it doesn’t matter, but see #390 / #409 for other users who care.)

Reply:
I don't think it would make a difference.